source/common/http/http1/codec_impl.cc

```diff
@@ -453,14 +457,24 @@ Status RequestEncoderImpl::encodeHeaders(const RequestHeaderMap& headers, bool e
     upgrade_request_ = true;
   }
 
-  connection_.copyToBuffer(method->value().getStringView().data(), method->value().size());
-  connection_.addCharToBuffer(' ');
+  auto& conn_buffer_helper = connection_.bufferHelper();
+
+  const uint64_t basic_header_size = method->value().size() + (host ? host->value().size() : 0) +
+                                     (path ? path->value().size() : 0);
+
+  conn_buffer_helper.reserveBuffer(4096 + basic_header_size);
```

Comment:
Is it somewhere to ensure the size of `headers` is less than 4096 bytes?

Reply:
We need not to ensure it. We always will reserve enough space for encoding. 

```
void StreamEncoderImpl::encodeHeader(absl::string_view key, absl::string_view value) {
  ASSERT(!key.empty());

  const uint64_t header_size = key.size() + value.size() + 4;

  auto& conn_buffer_helper = connection_.bufferHelper();
  conn_buffer_helper.reserveBuffer(header_size);

  conn_buffer_helper.writeToBuffer(key);
  conn_buffer_helper.writeToBuffer(':');
  conn_buffer_helper.writeToBuffer(' ');
  conn_buffer_helper.writeToBuffer(value);
  conn_buffer_helper.writeToBuffer(CRLF);

  bytes_meter_->addHeaderBytesSent(header_size);
}

```